### PR TITLE
Cli, add websocket component

### PR DESCRIPTION
**Pull request by mruz**

### CODE REVIEW

File: ice/cli/websocket/client.zep [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+        for name, value in headers {
+            if strncmp(name, "HTTP_", 5) === 0 {
+                let name = substr(name, 5);
+            }
+
+            if name !== false {
+                let name = str_replace("_", " ", strtolower(name)),
+                    name = str_replace("-", " ", strtolower(name)),
+                    name = str_replace(" ", "-", ucwords(name)),
```
Comment (Yahasana) [inline ice/cli/websocket/client.zep]: let name = ucwords(str_replace(
                         ["_", "-", " "], 
                         [" ", " ", "-"], 
                         strtolower(name)
                     ),  "-"),

File: ice/cli/websocket/server.zep [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+        let addr = parse_url(address);
+
+        if addr === false || !isset addr["scheme"] || !isset addr["host"] || !isset addr["port"] {
+            throw new Exception("Invalid address");
+        }
+
+        let context = stream_context_create();
+
+        if count(options) {
```
Comment (Yahasana) [inline ice/cli/websocket/server.zep:43]: if options {

File: ice/cli/websocket/client.zep [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        int i = 0;
+
+        let chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"$&/()=[]{}0123456789",
+            key = "",
+            length = strlen(chars);
+
+        while i < 16 {
+            let key .= chars[mt_rand(0, length - 1)],
+                i++;
+        }
```
Comment (Yahasana) [inline ice/cli/websocket/client.zep:105]: length = strlen(chars) - 1;
        while i < 16 {
            let key .= chars[mt_rand(0, length)],
                i++;
        }

File: ice/cli/websocket/client.zep [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        array_unshift(headers, "GET " . res . " HTTP/1.1");
+        this->sendClear(this->socket, implode("\r\n", headers)."\r\n");
+
+        let data = this->receiveClear(this->socket);
+
+        if !preg_match("(Sec-Websocket-Accept:\s*(.*)$)mUi", data, matches) {
+            throw new Exception("Bad response");
+        }
+
+        if trim(matches[1]) !== base64_encode(pack("H*", sha1(key.self::magic))) {
```
Comment (Yahasana) [inline ice/cli/websocket/client.zep]: replace all `self::` with `static::`

File: ice/cli/websocket/client.zep [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                this->normalizeHeaders(headers)
+            );
+
+        let key = headers["Sec-Websocket-Key"];
+
+        for name, value in headers {
+            let headers[name] = name . ": " . value;
+        }
+
+        if isset addr["path"] && strlen(addr["path"]) {
```
Comment (Yahasana) [inline ice/cli/websocket/client.zep]: why not `!empty addr["path"]`

File: ice/cli/websocket/client.zep [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            let res = addr["path"] . (isset addr["query"] && strlen(addr["query"]) ? "?" . addr["query"] : "");
```
Comment (Yahasana) [inline ice/cli/websocket/client.zep]: let res = addr["path"] . (empty addr["query"] ? "" : "?" . addr["query"]);

File: ice/cli/websocket/client.zep [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+            if stream_select(changed, null, null, (isset this->tick ? 0 : null)) > 0 {
+                for socket in changed {
+                    let message = this->receive(socket);
+
+                    if message !== false && isset this->message {
+                        call_user_func(this->message, message, this);
+                    }
+                }
+            }
+            usleep(5000);
```
Comment (Yahasana) [inline ice/cli/websocket/client.zep]: maybe we need a `waitingTime` property

File: ice/cli/websocket/server.zep (hunk 2 above)
Comment (mruz) [reply ice/cli/websocket/server.zep:43]: I can't do that in zephir _Variable can't be evaluated array_.

File: ice/cli/websocket/client.zep (hunk 5 above)
Comment (mruz) [reply ice/cli/websocket/client.zep]: I just noticed `empty` doesn't work like in PHP https://github.com/phalcon/zephir/issues/917.

**Pull request merged**